Provider crashes on "octopusdeploy_machine_policy" creation

When creating/updating `octopusdeploy_machine_policy` resource with any configuration block specified, provider crashes.

**Steps to reproduce**
1. Create a simple resource definition:

```
resource "octopusdeploy_machine_policy" "kubernetes" {
  name     = "Kubernetes"
  space_id = local.space_id

  connection_connect_timeout      = 5000000000
  connection_retry_count_limit    = 5
  connection_retry_sleep_interval = 20000000000
  connection_retry_time_limit     = 180000000000

  machine_connectivity_policy {
    machine_connectivity_behavior = "ExpectedToBeOnline"
  }
  machine_cleanup_policy {
    delete_machines_behavior         = "DeleteUnavailableMachines"
    delete_machines_elapsed_timespan = 3600000000000
  }
  machine_health_check_policy {
    health_check_type     = "OnlyConnectivity"
    health_check_interval = 15000000000
  }
  machine_update_policy {
    calamari_update_behavior = "UpdateAlways"
    tentacle_update_behavior = "NeverUpdate"
  }
}
```

2. Run `terraform apply`
3. Receive the following error:

```
╷
│ Error: Plugin did not respond
│ 
│   with octopusdeploy_machine_policy.kubernetes,
│   on infrastructure.tf line 1, in resource "octopusdeploy_machine_policy" "kubernetes":
│    1: resource "octopusdeploy_machine_policy" "kubernetes" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-octopusdeploy_v0.7.55 plugin:

panic: interface conversion: interface {} is *schema.Set, not []interface {}

goroutine 98 [running]:
github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy.expandMachineCleanupPolicy(...)
        github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy/schema_machine_cleanup_policy.go:12
github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy.expandMachinePolicy(0xc000049980, 0x23)
        github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy/schema_machine_policy.go:43 +0xcb7
github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy.resourceMachinePolicyUpdate(0x1ba3c28, 0xc000b8a600, 0xc000049980, 0x19030a0, 0xc000246280, 0xc000d8b6a0, 0x15125aa, 0xc00035f300)
        github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy/resource_machine_policy.go:86 +0xf6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc000461a40, 0x1ba3bb8, 0xc00011ae00, 0xc000049980, 0x19030a0, 0xc000246280, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:364 +0x17f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000461a40, 0x1ba3bb8, 0xc00011ae00, 0xc00013f420, 0xc00035f300, 0x19030a0, 0xc000246280, 0x0, 0x0, 0x0, ...)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00011c000, 0x1ba3bb8, 0xc00011ae00, 0xc0009336d0, 0xc00011ae00, 0x1a30fe0, 0xc00010ef00)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc000315960, 0x1ba3c60, 0xc00011ae00, 0xc00013f260, 0xc000315960, 0xc00010ef00, 0xc000a5dba0)
        github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1a30fe0, 0xc000315960, 0x1ba3c60, 0xc00010ef00, 0xc000aa8ea0, 0x0, 0x1ba3c60, 0xc00010ef00, 0xc00054e000, 0x963)
        github.com/hashicorp/terraform-plugin-go@v0.3.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000134380, 0x1baca18, 0xc000103c80, 0xc0005af560, 0xc000502b70, 0x20bfd60, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.39.1/server.go:1293 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc000134380, 0x1baca18, 0xc000103c80, 0xc0005af560, 0x0)
        google.golang.org/grpc@v1.39.1/server.go:1618 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000412190, 0xc000134380, 0x1baca18, 0xc000103c80, 0xc0005af560)
        google.golang.org/grpc@v1.39.1/server.go:941 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.39.1/server.go:939 +0x1fd

Error: The terraform-provider-octopusdeploy_v0.7.55 plugin crashed!
```

**Expected behavior**
Provider should not crash and successfully create resource.

**Logs and other supporting information**
If I omit all the attributes except for `name` and `space_id`, policy get's created with default values (I can see them in state and Octopus UI), but attempt to modify any of them (e.g. `connection_connect_timeout`) leads to the same error.

**Environment and versions:**
- OS: `MacOS 11.3.1`
- Octopus Server Version: `2021.1.7665`
- Terraform Version: `v1.0.4`
- Octopus Terraform Provider Version: `0.7.55`

--- a/octopusdeploy/resource_machine_policy_test.go
+++ b/octopusdeploy/resource_machine_policy_test.go
@@ -33,6 +33,22 @@ func TestAccMachinePolicyImportBasic(t *testing.T) {
 	})
 }
 
+func TestAccMachinePolicyIssue230(t *testing.T) {
+	localName := acctest.RandStringFromCharSet(20, acctest.CharSetAlpha)
+	name := acctest.RandStringFromCharSet(20, acctest.CharSetAlpha)
+
+	resource.Test(t, resource.TestCase{
+		CheckDestroy: testAccMachinePolicyCheckDestroy,
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testMachinePolicyIssue230(localName, name),
+			},
+		},
+	})
+}
+
 func TestAccMachinePolicyBasic(t *testing.T) {
 	localName := acctest.RandStringFromCharSet(20, acctest.CharSetAlpha)
 	prefix := "octopusdeploy_machine_policy." + localName
@@ -72,6 +88,39 @@ func testMachinePolicyDataSource(localName string, name string) string {
 func testMachinePolicyBasic(localName string, name string) string {
 	return fmt.Sprintf(`resource "octopusdeploy_machine_policy" "%s" {
 		name = "%s"
+	}`, localName, name)
+}
+
+func testMachinePolicyIssue230(localName string, name string) string {
+	return fmt.Sprintf(`resource "octopusdeploy_machine_policy" "%s" {
+		name = "%s"
+
+		machine_connectivity_policy {
+		  machine_connectivity_behavior = "ExpectedToBeOnline"
+		}
+
+		machine_cleanup_policy {
+		  delete_machines_behavior         = "DeleteUnavailableMachines"
+		  delete_machines_elapsed_timespan = 3600000000000
+		}
+
+		machine_health_check_policy {
+		  health_check_type     = "OnlyConnectivity"
+		  health_check_interval = 15000000000
+
+		  bash_health_check_policy {
+			run_type = "InheritFromDefault"
+		  }
+
+		  powershell_health_check_policy {
+			run_type = "InheritFromDefault"
+		}
+	  }
+
+		machine_update_policy {
+		  calamari_update_behavior = "UpdateAlways"
+		  tentacle_update_behavior = "NeverUpdate"
+		}
 	}`, localName, name)
 }
 

--- a/octopusdeploy/schema_machine_cleanup_policy.go
+++ b/octopusdeploy/schema_machine_cleanup_policy.go
@@ -9,13 +9,27 @@ import (
 )
 
 func expandMachineCleanupPolicy(values interface{}) *octopusdeploy.MachineCleanupPolicy {
-	flattenedValues := values.([]interface{})
-	flattenedMap := flattenedValues[0].(map[string]interface{})
-
-	return &octopusdeploy.MachineCleanupPolicy{
-		DeleteMachinesBehavior:        flattenedMap["delete_machines_behavior"].(string),
-		DeleteMachinesElapsedTimeSpan: time.Duration(flattenedMap["delete_machines_elapsed_timespan"].(int)),
+	if values == nil {
+		return nil
 	}
+	flattenedValues := values.(*schema.Set)
+	if len(flattenedValues.List()) == 0 {
+		return nil
+	}
+
+	flattenedMap := flattenedValues.List()[0].(map[string]interface{})
+
+	machineCleanupPolicy := octopusdeploy.NewMachineCleanupPolicy()
+
+	if v, ok := flattenedMap["delete_machines_behavior"]; ok {
+		machineCleanupPolicy.DeleteMachinesBehavior = v.(string)
+	}
+
+	if v, ok := flattenedMap["delete_machines_elapsed_timespan"]; ok {
+		machineCleanupPolicy.DeleteMachinesElapsedTimeSpan = time.Duration(v.(int))
+	}
+
+	return machineCleanupPolicy
 }
 
 func flattenMachineCleanupPolicy(machineCleanupPolicy *octopusdeploy.MachineCleanupPolicy) []interface{} {

--- a/octopusdeploy/schema_machine_connectivity_policy.go
+++ b/octopusdeploy/schema_machine_connectivity_policy.go
@@ -7,12 +7,23 @@ import (
 )
 
 func expandMachineConnectivityPolicy(values interface{}) *octopusdeploy.MachineConnectivityPolicy {
-	flattenedValues := values.([]interface{})
-	flattenedMap := flattenedValues[0].(map[string]interface{})
-
-	return &octopusdeploy.MachineConnectivityPolicy{
-		MachineConnectivityBehavior: flattenedMap["machine_connectivity_behavior"].(string),
+	if values == nil {
+		return nil
 	}
+	flattenedValues := values.(*schema.Set)
+	if len(flattenedValues.List()) == 0 {
+		return nil
+	}
+
+	flattenedMap := flattenedValues.List()[0].(map[string]interface{})
+
+	machineConnectivityPolicy := octopusdeploy.NewMachineConnectivityPolicy()
+
+	if v, ok := flattenedMap["machine_connectivity_behavior"]; ok {
+		machineConnectivityPolicy.MachineConnectivityBehavior = v.(string)
+	}
+
+	return machineConnectivityPolicy
 }
 
 func flattenMachineConnectivityPolicy(machineConnectivityPolicy *octopusdeploy.MachineConnectivityPolicy) []interface{} {

--- a/octopusdeploy/schema_machine_health_check_policy.go
+++ b/octopusdeploy/schema_machine_health_check_policy.go
@@ -9,19 +9,49 @@ import (
 )
 
 func expandMachineHealthCheckPolicy(values interface{}) *octopusdeploy.MachineHealthCheckPolicy {
-	flattenedValues := values.([]interface{})
-	flattenedMap := flattenedValues[0].(map[string]interface{})
-
-	duration, _ := time.ParseDuration(flattenedMap["health_check_interval"].(string))
-
-	return &octopusdeploy.MachineHealthCheckPolicy{
-		BashHealthCheckPolicy:       expandMachineScriptPolicy(flattenedMap["bash_health_check_policy"]),
-		HealthCheckCron:             flattenedMap["health_check_cron"].(string),
-		HealthCheckCronTimezone:     flattenedMap["health_check_cron_timezone"].(string),
-		HealthCheckInterval:         duration,
-		HealthCheckType:             flattenedMap["health_check_type"].(string),
-		PowerShellHealthCheckPolicy: expandMachineScriptPolicy(flattenedMap["powershell_health_check_policy"]),
+	if values == nil {
+		return nil
 	}
+	flattenedValues := values.(*schema.Set)
+	if len(flattenedValues.List()) == 0 {
+		return nil
+	}
+
+	flattenedMap := flattenedValues.List()[0].(map[string]interface{})
+
+	machineHealthCheckPolicy := octopusdeploy.NewMachineHealthCheckPolicy()
+
+	if v, ok := flattenedMap["bash_health_check_policy"]; ok {
+		if len(v.([]interface{})) > 0 {
+			machineHealthCheckPolicy.BashHealthCheckPolicy = expandMachineScriptPolicy(v)
+		}
+	}
+
+	if v, ok := flattenedMap["health_check_cron"]; ok {
+		machineHealthCheckPolicy.HealthCheckCron = v.(string)
+	}
+
+	if v, ok := flattenedMap["health_check_cron_timezone"]; ok {
+		if s := v.(string); len(s) > 0 {
+			machineHealthCheckPolicy.HealthCheckCronTimezone = s
+		}
+	}
+
+	if v, ok := flattenedMap["health_check_interval"]; ok {
+		machineHealthCheckPolicy.HealthCheckInterval = time.Duration(v.(int))
+	}
+
+	if v, ok := flattenedMap["health_check_type"]; ok {
+		machineHealthCheckPolicy.HealthCheckType = v.(string)
+	}
+
+	if v, ok := flattenedMap["powershell_health_check_policy"]; ok {
+		if len(v.([]interface{})) > 0 {
+			machineHealthCheckPolicy.PowerShellHealthCheckPolicy = expandMachineScriptPolicy(v)
+		}
+	}
+
+	return machineHealthCheckPolicy
 }
 
 func flattenMachineHealthCheckPolicy(machineHealthCheckPolicy *octopusdeploy.MachineHealthCheckPolicy) []interface{} {
@@ -44,7 +74,7 @@ func getMachineHealthCheckPolicySchema() map[string]*schema.Schema {
 		"bash_health_check_policy": {
 			Elem:     &schema.Resource{Schema: getMachineScriptPolicySchema()},
 			MaxItems: 1,
-			Optional: true,
+			Required: true,
 			Type:     schema.TypeList,
 		},
 		"health_check_cron": {
@@ -52,10 +82,12 @@ func getMachineHealthCheckPolicySchema() map[string]*schema.Schema {
 			Type:     schema.TypeString,
 		},
 		"health_check_cron_timezone": {
+			Default:  "UTC",
 			Optional: true,
 			Type:     schema.TypeString,
 		},
 		"health_check_interval": {
+			Default:  24 * time.Hour,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
@@ -71,7 +103,7 @@ func getMachineHealthCheckPolicySchema() map[string]*schema.Schema {
 		"powershell_health_check_policy": {
 			Elem:     &schema.Resource{Schema: getMachineScriptPolicySchema()},
 			MaxItems: 1,
-			Optional: true,
+			Required: true,
 			Type:     schema.TypeList,
 		},
 	}

--- a/octopusdeploy/schema_machine_policy.go
+++ b/octopusdeploy/schema_machine_policy.go
@@ -40,19 +40,27 @@ func expandMachinePolicy(d *schema.ResourceData) *octopusdeploy.MachinePolicy {
 	}
 
 	if v, ok := d.GetOk("machine_cleanup_policy"); ok {
-		machinePolicy.MachineCleanupPolicy = expandMachineCleanupPolicy(v)
+		if len(v.(*schema.Set).List()) > 0 {
+			machinePolicy.MachineCleanupPolicy = expandMachineCleanupPolicy(v)
+		}
 	}
 
 	if v, ok := d.GetOk("machine_connectivity_policy"); ok {
-		machinePolicy.MachineConnectivityPolicy = expandMachineConnectivityPolicy(v)
+		if len(v.(*schema.Set).List()) > 0 {
+			machinePolicy.MachineConnectivityPolicy = expandMachineConnectivityPolicy(v)
+		}
 	}
 
 	if v, ok := d.GetOk("machine_health_check_policy"); ok {
-		machinePolicy.MachineHealthCheckPolicy = expandMachineHealthCheckPolicy(v)
+		if len(v.(*schema.Set).List()) > 0 {
+			machinePolicy.MachineHealthCheckPolicy = expandMachineHealthCheckPolicy(v)
+		}
 	}
 
 	if v, ok := d.GetOk("machine_update_policy"); ok {
-		machinePolicy.MachineUpdatePolicy = expandMachineUpdatePolicy(v)
+		if len(v.(*schema.Set).List()) > 0 {
+			machinePolicy.MachineUpdatePolicy = expandMachineUpdatePolicy(v)
+		}
 	}
 
 	if v, ok := d.GetOk("name"); ok {
@@ -64,7 +72,7 @@ func expandMachinePolicy(d *schema.ResourceData) *octopusdeploy.MachinePolicy {
 	}
 
 	if v, ok := d.GetOk("polling_request_queue_timeout"); ok {
-		machinePolicy.PollingRequestMaximumMessageProcessingTimeout = time.Duration(v.(int))
+		machinePolicy.PollingRequestQueueTimeout = time.Duration(v.(int))
 	}
 
 	if v, ok := d.GetOk("space_id"); ok {
@@ -94,7 +102,7 @@ func flattenMachinePolicy(machinePolicy *octopusdeploy.MachinePolicy) map[string
 		"machine_update_policy":           flattenMachineUpdatePolicy(machinePolicy.MachineUpdatePolicy),
 		"name":                            machinePolicy.Name,
 		"polling_request_maximum_message_processing_timeout": machinePolicy.PollingRequestMaximumMessageProcessingTimeout,
-		"polling_request_queue_timeout":                      machinePolicy.PollingRequestMaximumMessageProcessingTimeout,
+		"polling_request_queue_timeout":                      machinePolicy.PollingRequestQueueTimeout,
 		"space_id":                                           machinePolicy.SpaceID,
 	}
 }
@@ -121,22 +129,22 @@ func getMachinePolicyDataSchema() map[string]*schema.Schema {
 func getMachinePolicySchema() map[string]*schema.Schema {
 	return map[string]*schema.Schema{
 		"connection_connect_timeout": {
-			Computed: true,
+			Default:  time.Minute,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
 		"connection_retry_count_limit": {
-			Computed: true,
+			Default:  5,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
 		"connection_retry_sleep_interval": {
-			Computed: true,
+			Default:  time.Second,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
 		"connection_retry_time_limit": {
-			Computed: true,
+			Default:  5 * time.Minute,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
@@ -147,28 +155,24 @@ func getMachinePolicySchema() map[string]*schema.Schema {
 			Type:     schema.TypeBool,
 		},
 		"machine_cleanup_policy": {
-			Computed: true,
 			Elem:     &schema.Resource{Schema: getMachineCleanupPolicySchema()},
 			MaxItems: 1,
 			Optional: true,
 			Type:     schema.TypeSet,
 		},
 		"machine_connectivity_policy": {
-			Computed: true,
 			Elem:     &schema.Resource{Schema: getMachineConnectivityPolicySchema()},
 			MaxItems: 1,
 			Optional: true,
 			Type:     schema.TypeSet,
 		},
 		"machine_health_check_policy": {
-			Computed: true,
 			Elem:     &schema.Resource{Schema: getMachineHealthCheckPolicySchema()},
 			MaxItems: 1,
 			Optional: true,
 			Type:     schema.TypeSet,
 		},
 		"machine_update_policy": {
-			Computed: true,
 			Elem:     &schema.Resource{Schema: getMachineUpdatePolicySchema()},
 			MaxItems: 1,
 			Optional: true,
@@ -176,12 +180,12 @@ func getMachinePolicySchema() map[string]*schema.Schema {
 		},
 		"name": getNameSchema(true),
 		"polling_request_maximum_message_processing_timeout": {
-			Computed: true,
+			Default:  10 * time.Minute,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
 		"polling_request_queue_timeout": {
-			Computed: true,
+			Default:  2 * time.Minute,
 			Optional: true,
 			Type:     schema.TypeInt,
 		},
@@ -197,6 +201,10 @@ func setMachinePolicy(ctx context.Context, d *schema.ResourceData, machinePolicy
 	d.Set("description", machinePolicy.Description)
 	d.Set("id", machinePolicy.GetID())
 	d.Set("is_default", machinePolicy.IsDefault)
+	d.Set("name", machinePolicy.Name)
+	d.Set("polling_request_maximum_message_processing_timeout", machinePolicy.PollingRequestMaximumMessageProcessingTimeout)
+	d.Set("polling_request_queue_timeout", machinePolicy.PollingRequestQueueTimeout)
+	d.Set("space_id", machinePolicy.SpaceID)
 
 	if err := d.Set("machine_cleanup_policy", flattenMachineCleanupPolicy(machinePolicy.MachineCleanupPolicy)); err != nil {
 		return fmt.Errorf("error setting machine_cleanup_policy: %s", err)
@@ -213,11 +221,6 @@ func setMachinePolicy(ctx context.Context, d *schema.ResourceData, machinePolicy
 	if err := d.Set("machine_update_policy", flattenMachineUpdatePolicy(machinePolicy.MachineUpdatePolicy)); err != nil {
 		return fmt.Errorf("error setting machine_update_policy: %s", err)
 	}
-
-	d.Set("name", machinePolicy.Name)
-	d.Set("polling_request_maximum_message_processing_timeout", machinePolicy.PollingRequestMaximumMessageProcessingTimeout)
-	d.Set("polling_request_queue_timeout", machinePolicy.PollingRequestMaximumMessageProcessingTimeout)
-	d.Set("space_id", machinePolicy.SpaceID)
 
 	return nil
 }

--- a/octopusdeploy/schema_machine_script_policy.go
+++ b/octopusdeploy/schema_machine_script_policy.go
@@ -3,16 +3,32 @@ package octopusdeploy
 import (
 	"github.com/OctopusDeploy/go-octopusdeploy/octopusdeploy"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 )
 
 func expandMachineScriptPolicy(values interface{}) *octopusdeploy.MachineScriptPolicy {
+	if values == nil {
+		return nil
+	}
 	flattenedValues := values.([]interface{})
+	if len(flattenedValues) == 0 {
+		return nil
+	}
+
 	flattenedMap := flattenedValues[0].(map[string]interface{})
 
-	return &octopusdeploy.MachineScriptPolicy{
-		RunType:    flattenedMap["run_type"].(string),
-		ScriptBody: flattenedMap["script_body"].(*string),
+	machineScriptPolicy := octopusdeploy.NewMachineScriptPolicy()
+
+	if v, ok := flattenedMap["run_type"]; ok {
+		machineScriptPolicy.RunType = v.(string)
 	}
+
+	if v, ok := flattenedMap["script_body"]; ok {
+		scriptBody := v.(string)
+		machineScriptPolicy.ScriptBody = &scriptBody
+	}
+
+	return machineScriptPolicy
 }
 
 func flattenMachineScriptPolicy(machineScriptPolicy *octopusdeploy.MachineScriptPolicy) []interface{} {
@@ -29,8 +45,14 @@ func flattenMachineScriptPolicy(machineScriptPolicy *octopusdeploy.MachineScript
 func getMachineScriptPolicySchema() map[string]*schema.Schema {
 	return map[string]*schema.Schema{
 		"run_type": {
+			Default:  "InheritFromDefault",
 			Optional: true,
 			Type:     schema.TypeString,
+			ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{
+				"InheritFromDefault",
+				"Inline",
+				"OnlyConnectivity",
+			}, false)),
 		},
 		"script_body": {
 			Optional: true,

--- a/octopusdeploy/schema_machine_update_policy.go
+++ b/octopusdeploy/schema_machine_update_policy.go
@@ -7,14 +7,31 @@ import (
 )
 
 func expandMachineUpdatePolicy(values interface{}) *octopusdeploy.MachineUpdatePolicy {
-	flattenedValues := values.([]interface{})
-	flattenedMap := flattenedValues[0].(map[string]interface{})
-
-	return &octopusdeploy.MachineUpdatePolicy{
-		CalamariUpdateBehavior:  flattenedMap["calamari_update_behavior"].(string),
-		TentacleUpdateAccountID: flattenedMap["tentacle_update_account_id"].(string),
-		TentacleUpdateBehavior:  flattenedMap["tentacle_update_behavior"].(string),
+	if values == nil {
+		return nil
 	}
+	flattenedValues := values.(*schema.Set)
+	if len(flattenedValues.List()) == 0 {
+		return nil
+	}
+
+	flattenedMap := flattenedValues.List()[0].(map[string]interface{})
+
+	machineUpdatePolicy := octopusdeploy.NewMachineUpdatePolicy()
+
+	if v, ok := flattenedMap["calamari_update_behavior"]; ok {
+		machineUpdatePolicy.CalamariUpdateBehavior = v.(string)
+	}
+
+	if v, ok := flattenedMap["tentacle_update_account_id"]; ok {
+		machineUpdatePolicy.TentacleUpdateAccountID = v.(string)
+	}
+
+	if v, ok := flattenedMap["tentacle_update_behavior"]; ok {
+		machineUpdatePolicy.TentacleUpdateBehavior = v.(string)
+	}
+
+	return machineUpdatePolicy
 }
 
 func flattenMachineUpdatePolicy(machineUpdatePolicy *octopusdeploy.MachineUpdatePolicy) []interface{} {

